tests/js/server/replication2/replication2-replicated-state-document-store.js

```diff
@@ -578,6 +586,255 @@ const replicatedStateRecoverySuite = function () {
+      // Checks that inserted documents are still there after compaction is done.
+      // Various errors that could occur during recovery are triggered intentionally (e.g. unique constraint violation).
+      // Because after the leader election followers also replay the un-compacted part of the log, this test also covers
+      // error handling on followers.
+      let documents = [];
+      for (let counter = 0; counter < 100; ++counter) {
+        documents.push({_key: `${testName}-${counter}`, value: counter, temporary: counter});
+      }
+      for (let idx = 0; idx < documents.length; ++idx) {
+        collection.insert(documents[idx]);
+      }
+
+      // The following operation will be used to trigger errors, which should be ignored during recovery.
+      const removedDoc = {_key: `${testName}-document-not-found`};
+      collection.insert(removedDoc);
+
+      // Wait for operations to be synced
+      let {leader} = lh.getReplicatedLogLeaderPlan(database, logId);
+      let status = log.status();
+      const commitIndex = status.participants[leader].response.local.commitIndex;
+      lh.waitFor(lp.lowestIndexToKeepReached(log, leader, commitIndex));
+
+      // Trigger compaction intentionally.
+      log.compact();
+
+      // Should ignore document-not-found
+      collection.remove(removedDoc);
+
+      // Create a temporary unique index on temporary.
+      let uniqueIndex = collection.ensureIndex({name: "kuh", type: "persistent", fields: ["temporary"], unique: true});
+      assertEqual(uniqueIndex.name, "kuh");
+      // Drop the index
+      assertTrue(collection.dropIndex(uniqueIndex));
+      // Wait for the index to removed from Current. This is how we know that the leader dropped the index.
+      let indexId = uniqueIndex.id.split('/')[1];
+      lh.waitFor(() => {
+        if (dh.isIndexInCurrent(database, collection._id, indexId)) {
+          return Error(`Index ${indexId} still in Current`);
+        }
+        return true;
+      });
+
+      // Insert a doc that has the same temporary as an existing one.
+      // When the index creation is going to be replayed, it will fail, but the error should be ignored.
+      collection.insert({_key: `${testName}-temporary`, value: 1234567, temporary: 0})
+
+      // Insert two docs that have a common property
+      const commonDoc1 = {_key: `${testName}-common1`, value: 1234};
+      const commonDoc2 = {_key: `${testName}-common2`, value: 1234};
+      collection.insert(commonDoc1);
+      collection.insert(commonDoc2);
+      // Delete the first doc
+      collection.remove(commonDoc1);
+      // Create an unique index. When commonDoc1 insertions is replayed, it should trigger unique constraint violation.
+      // The error should be ignored during recovery.
+      uniqueIndex = collection.ensureIndex({name: "eule", type: "persistent", fields: ["value"], unique: true});
+      assertEqual(uniqueIndex.name, "eule");
+
+      // Trigger leader recovery
```

Comment:
Just for my understanding is it guaranteed that we do not do any leader compaction after the "compact" call above?
Otherwise we may not trigger the bad situation.

Reply:
In theory, no, there's no guarantee.
The framework is free do compact whenever it sees fit, which is out of the replicated state's control. However, compaction can be costly, so there's a lower bound of 1000 log entries that need to be accumulated in order for the framework to consider a compaction. This is far from what we have in this test, and that is the reason I had to call `log.compact()` in order to "force" a compaction with so few log entries.
So in practice, the test is guaranteed to trigger the bad situation. If there will ever be a change to the compaction limit such that it becomes ridiculously low, I can implement a work-around with `TRI_IF_FAILURE`, but that is overkill given the current situation.

Reply:
Ok agreed :+1: